arch/x86/core/thread.c

```diff
@@ -188,3 +143,95 @@ void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
 	thread_monitor_init(thread);
 #endif
 }
+
+#ifdef CONFIG_X86_USERSPACE
+void _x86_swap_update_page_tables(struct k_thread *incoming,
+				  struct k_thread *outgoing)
+{
+	/* Outgoing thread stack no longer accessible */
+	_x86_mmu_set_flags((void *)outgoing->stack_info.start,
+			   ROUND_UP(outgoing->stack_info.size, MMU_PAGE_SIZE),
+			   MMU_ENTRY_SUPERVISOR, MMU_PTE_US_MASK);
+
+
+	/* Userspace can now access the incoming thread's stack */
+	_x86_mmu_set_flags((void *)incoming->stack_info.start,
+			   ROUND_UP(incoming->stack_info.size, MMU_PAGE_SIZE),
+			   MMU_ENTRY_USER, MMU_PTE_US_MASK);
+
+	/* In case of privilege elevation, use the incoming thread's kernel
+	 * stack, the top of the thread stack is the bottom of the kernel stack
+	 */
+	_main_tss.esp0 = incoming->stack_info.start;
+
+	/* TODO: if either thread defines different memory domains, efficiently
+	 * switch between them
+	 */
+}
+
+
+int _arch_is_user_context(void)
+{
+	int cs;
+
+	/* On x86, read the CS register (which cannot be manually set)
+	 * and check the CPL bits within
+	 */
+	__asm__ volatile ("mov %%cs, %[cs_val]" : [cs_val] "=r" (cs));
+
+	/* Mask all but the lowest 2 bits, which is the CPL information */
+	cs &= 0x3;
+
+	/* Only 0 if in supervisor mode */
+	return cs;
+}
+
+
+FUNC_NORETURN void _arch_user_mode_enter(_thread_entry_t user_entry,
+					 void *p1, void *p2, void *p3)
+{
+	u32_t stack_end;
+
+	/* Transition will reset stack pointer to initial, discarding
+	 * any old context since this is a one-way operation
+	 */
+	stack_end = STACK_ROUND_DOWN(_current->stack_info.start +
+				     _current->stack_info.size);
+
+	/* Set up the kernel stack used during privilege elevation */
+	_x86_mmu_set_flags((void *)(_current->stack_info.start - MMU_PAGE_SIZE),
+			   MMU_PAGE_SIZE,
+			   (MMU_ENTRY_PRESENT | MMU_ENTRY_WRITE |
+			    MMU_ENTRY_SUPERVISOR),
+			   (MMU_PTE_P_MASK | MMU_PTE_RW_MASK |
+			    MMU_PTE_US_MASK));
+
```

Comment:
Same point as above: does a far call that changes the ring level do a TLB flush?  I'm not sure that it does.  Normal OSes will always rewrite CR3 on context switch, which will obviously flush the cache.  But this code looks like it's just changing page tables in place but not actually telling the CPU that the page tables have been changed.

Reply:
_x86_mmu_set_flags() flushes pages for the relevant memory addresses.

Reply:
Was looking into this today. Actually we can think of this if it improves the performance of our mmu. Instead of calling a cache flush for each page write we call it only once. So what we do is, just rewrite the same start address into CR3 when we finish all the page table manipulations.

Reply:
I considered this, but it's not clear what the tradeoffs are.
Resetting CR3 is just one instruction, but the CPU has to do a lot of work once that is done since it has to invalidate everything.
It's not clear what the threshold is, where resetting CR3 performs better than using "invlpg" on only those pages which were changed.

Some discussion here: https://stackoverflow.com/questions/28384234/when-to-do-or-not-do-invlpg-mov-to-cr3-to-minimize-tlb-flushing